azurerm/resource_arm_analysis_services_server.go

```diff
@@ -0,0 +1,449 @@
+	//}
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
+
+	if connectionMode := expandAnalysisServicesServerQuerypoolConnectionMode(d); connectionMode != nil {
+		serverProperties.QuerypoolConnectionMode = *connectionMode
+	}
+
+	return &serverProperties
+}
+
+func expandAnalysisServicesServerAdminUsers(d *schema.ResourceData) *analysisservices.ServerAdministrators {
+	adminUsers := d.Get("admin_users").(*schema.Set)
+	members := make([]string, 0)
+
+	for _, admin := range adminUsers.List() {
+		if adm, ok := admin.(string); ok {
+			members = append(members, adm)
+		}
+	}
+
+	return &analysisservices.ServerAdministrators{Members: &members}
+}
+
+func expandAnalysisServicesServerQuerypoolConnectionMode(d *schema.ResourceData) *analysisservices.ConnectionMode {
+	if querypoolConnectionMode, ok := d.GetOk("querypool_connection_mode"); ok {
+		connMode := analysisservices.ConnectionMode(querypoolConnectionMode.(string))
+		return &connMode
+	}
+
+	return nil
+}
+
+func expandAnalysisServicesServerFirewallSettings(d *schema.ResourceData) *analysisservices.IPv4FirewallSettings {
+	firewallSettings := analysisservices.IPv4FirewallSettings{}
+
+	if enablePowerBi, exists := d.GetOkExists("enable_power_bi_service"); exists {
+		firewallSettings.EnablePowerBIService = utils.Bool(enablePowerBi.(bool))
+	}
+
+	firewallRules := d.Get("ipv4_firewall_rule").([]interface{})
+
+	fwRules := make([]analysisservices.IPv4FirewallRule, len(firewallRules))
+	for i, v := range firewallRules {
+		fwRule := v.(map[string]interface{})
+		fwRules[i] = analysisservices.IPv4FirewallRule{
+			FirewallRuleName: utils.String(fwRule["name"].(string)),
+			RangeStart:       utils.String(fwRule["range_start"].(string)),
+			RangeEnd:         utils.String(fwRule["range_end"].(string)),
+		}
+	}
+	firewallSettings.FirewallRules = &fwRules
+
+	return &firewallSettings
+}
+
+func flattenAnalysisServicesServerAdminUsers(serverAdministrators *analysisservices.ServerAdministrators) *[]string {
+	if serverAdministrators == nil || serverAdministrators.Members == nil {
+		return &[]string{}
+	}
+
+	return serverAdministrators.Members
+}
+
+func flattenAnalysisServicesServerFirewallSettings(serverProperties *analysisservices.ServerProperties) (enablePowerBi *bool, fwRules []interface{}) {
+	if serverProperties.IPV4FirewallSettings == nil {
+		return nil, nil
+	}
+
+	firewallSettings := serverProperties.IPV4FirewallSettings
+
+	fwRules = make([]interface{}, 0)
+	for _, fwRule := range *firewallSettings.FirewallRules {
+		output := make(map[string]interface{})
+		output["name"] = *fwRule.FirewallRuleName
+		output["range_start"] = *fwRule.RangeStart
+		output["range_end"] = *fwRule.RangeEnd
+
+		fwRules = append(fwRules, output)
+	}
+
+	return firewallSettings.EnablePowerBIService, fwRules
+}
+
+func flattenAnalysisServicesServerQuerypoolConnectionMode(serverProperties *analysisservices.ServerProperties) *string {
```

Comment:
I don't think this function is necessary.

Reply:
removed (in case you mean we can do it inline)